File: Practice/atcoder/ABC/077/src/c.py
# それぞれソート済みとして, 整数でAi<BjかつBj<Ckとなるjを求める(にぶたん)

N = int(input())
A = list(sorted(list(map(int, input().split()))))
B = list(sorted(list(map(int, input().split()))))
C = list(sorted(list(map(int, input().split()))))


def lower_bound(arr, target):
    # targetを超える最小の値のインデックス
    if target > arr[-1]:
        return None
    left, right = 0, len(arr)
    while left < right:
        mid = (left + right) // 2
        if arr[mid] > target:
            right = mid
        else:
            left = mid + 1
    return right


def upper_bound(arr, target):
    # target以上の最小の値のインデックス
    if target < arr[0]:
        return None
    left, right = 0, len(arr)
    while left < right:
        mid = (left + right) // 2
        if arr[mid] >= target:
            right = mid
        else:
            left = mid + 1
    return right


count = 0

for b in B:
    ai = upper_bound(A, b)
    ci = lower_bound(C, b)
    if ai is not None and ci is not None:
        count += ai * (N - ci)

print(count)

File: Practice/atcoder/ABC/077/src/_c.php
<?php
  # にぶたんのいい問題らしいぞ もう以前どうやって解いたかとか忘れたけど
  $n = intVal(trim(fgets(STDIN)));
  $as = array_map("intVal", explode(" ", trim(fgets(STDIN))));
  $bs = array_map("intVal", explode(" ", trim(fgets(STDIN))));
  $cs = array_map("intVal", explode(" ", trim(fgets(STDIN))));
  sort($as);
  sort($bs);
  sort($cs);

  function lower_bound($t, $arr){
    $l = 0; 
    $r = count($arr);
    while($l < $r){
      $mid = floor(($l + $r) / 2);
      if($arr[$mid] > $t){
        $r = $mid;
      }else{
        $l = $mid + 1;
      }
    }
    return $r;
  }

  $ans = 0;
  foreach($as as $a){
    $b = lower_bound($a, $bs);
    $c = lower_bound($b, $cs);
    $ans += $b*$c;
  }
  echo $ans . PHP_EOL;
?>